modules/puppetca/puppetca_api.rb

```diff
@@ -29,13 +29,27 @@ class Api < ::Sinatra::Base
     post "/autosign/:certname" do
       content_type :json
       certname = params[:certname]
+      token_ttl = params[:token_ttl]
       begin
-        autosigner.autosign(certname)
+        autosigner.autosign(certname, token_ttl)
       rescue => e
         log_halt 406, "Failed to enable autosign for #{certname}: #{e}"
       end
     end
 
+    post "/validate" do
+      content_type :json
+      unless autosigner.respond_to?(:validate_csr)
+        log_halt 501, "Provider only supports trivial autosigning"
+      end
+      begin
+        request.body.rewind
+        autosigner.validate_csr(request.body.read) ? 200 : 404
```

Comment:
I'm wondering about the 404 code; will there be a need to distinguish between a proxy without this feature and one with? Probably not but it might be worth to consider 400 Bad Request for invalid CSRs.

Reply:
Well the 404 here only gets returned when the CSR is valid but the token is not. A proxy without this feature returns 501 (line 43) and an invalid CSR raises an Exception leading to a 406 (line 49). Does that work for you?